Entities/SLTerrain.cpp

```diff
@@ -1,1436 +1,474 @@
+			release_bitmap(fgTextureBitmap);
+		}
+		for (BITMAP *bgTextureBitmap : materialBGTextures) {
+			release_bitmap(bgTextureBitmap);
+		}
+		*/
 	}
 
-	// Register terrain change
-	g_SceneMan.RegisterTerrainChange(loc.m_X, loc.m_Y, pTObject->GetBitmapWidth(), pTObject->GetBitmapHeight(), g_MaskColor, false);
-}
-
-
-//////////////////////////////////////////////////////////////////////////////////////////
-// Virtual method:  ApplyTerrainObject
-//////////////////////////////////////////////////////////////////////////////////////////
-// Description:     Draws a passed in TerrainObject's graphical and material
-//                  representations to this Terrain's respective layers.
-
-void SLTerrain::ApplyTerrainObject(TerrainObject *pTObject)
-{
-    if (!pTObject)
-        return;
-
-    Vector loc = pTObject->GetPos() + pTObject->GetBitmapOffset();
-
-    // Do duplicate drawing if the terrain object straddles a wrapping border
-    if (loc.m_X < 0)
-    {
-        draw_sprite(m_pMainBitmap, pTObject->GetMaterialBitmap(), loc.m_X + m_pMainBitmap->w, loc.m_Y);
-        draw_sprite(m_pFGColor->GetBitmap(), pTObject->GetFGColorBitmap(), loc.m_X + m_pFGColor->GetBitmap()->w, loc.m_Y);
-        if (pTObject->HasBGColor())
-            draw_sprite(m_pBGColor->GetBitmap(), pTObject->GetBGColorBitmap(), loc.m_X + m_pBGColor->GetBitmap()->w, loc.m_Y);
-    }
-    else if (loc.m_X >= m_pMainBitmap->w - pTObject->GetFGColorBitmap()->w)
-    {
-        draw_sprite(m_pMainBitmap, pTObject->GetMaterialBitmap(), loc.m_X - m_pMainBitmap->w, loc.m_Y);
-        draw_sprite(m_pFGColor->GetBitmap(), pTObject->GetFGColorBitmap(), loc.m_X - m_pFGColor->GetBitmap()->w, loc.m_Y);
-        if (pTObject->HasBGColor())
-            draw_sprite(m_pBGColor->GetBitmap(), pTObject->GetBGColorBitmap(), loc.m_X - m_pBGColor->GetBitmap()->w, loc.m_Y);
-    }
-
-    // Regular drawing
-    draw_sprite(m_pMainBitmap, pTObject->GetMaterialBitmap(), loc.m_X, loc.m_Y);
-    draw_sprite(m_pFGColor->GetBitmap(), pTObject->GetFGColorBitmap(), loc.m_X, loc.m_Y);
-	if (pTObject->HasBGColor())
-	{
-		draw_sprite(m_pBGColor->GetBitmap(), pTObject->GetBGColorBitmap(), loc.m_X, loc.m_Y);
-		g_SceneMan.RegisterTerrainChange(loc.m_X, loc.m_Y, pTObject->GetBitmapWidth(), pTObject->GetBitmapHeight(), g_MaskColor, true);
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	int SLTerrain::LoadData() {
+		// Load the materials bitmap into the main bitmap.
+		SceneLayer::LoadData();
+
+		RTEAssert(m_FGColorLayer.get(), "Terrain's foreground layer not instantiated before trying to load its data!");
+		RTEAssert(m_BGColorLayer.get(), "Terrain's background layer not instantiated before trying to load its data!");
+
+		if (m_FGColorLayer->IsLoadedFromDisk() && m_BGColorLayer->IsLoadedFromDisk()) {
+			m_FGColorLayer->LoadData();
+			m_BGColorLayer->LoadData();
+		} else {
```

Comment:
Wouldn't make sense. FG/BG terrain layers are only loaded from disk if they were dumped for a savegame, so lading part of the layers just ends up with a mess.
It's not exactly a foolproof system because it just relies on data paths, so you could just provide both FG/BG layers to any random terrain preset and load everything from disk, but it's kinda pointless overall.